

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Tutorial &mdash; obogaf::parser  documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script type="text/javascript" src="_static/jquery.js"></script>
        <script type="text/javascript" src="_static/underscore.js"></script>
        <script type="text/javascript" src="_static/doctools.js"></script>
        <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Frequently Asked Questions" href="faq.html" />
    <link rel="prev" title="Usage of obogaf::parser" href="usage.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> obogaf::parser
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Installation &amp; Getting Started</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="quickstart.html">Quickstart</a></li>
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation</a></li>
</ul>
<p class="caption"><span class="caption-text">Usage &amp; Tutorial</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="usage.html">Usage of obogaf::parser</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Tutorial</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#application-to-the-gene-ontology-go">Application to the Gene Ontology (GO)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#parse-the-go-obo-file">Parse the GO obo file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#parse-the-goa-annotation-file">Parse the GOA annotation file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#map-go-terms-between-releases">Map GO terms between releases</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#application-to-human-phenotype-ontology-hpo">Application to Human Phenotype Ontology (HPO)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#parse-the-hpo-obo-file">Parse the HPO obo file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#parse-the-hpo-annotation-file">Parse the HPO annotation file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#map-hpo-terms-between-releases">Map HPO terms between releases</a></li>
</ul>
</li>
</ul>
</li>
</ul>
<p class="caption"><span class="caption-text">Tips &amp; Tricks</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="faq.html">Frequently Asked Questions</a></li>
</ul>
<p class="caption"><span class="caption-text">Project Info</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="citing.html">Cite obogaf::parser</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributing</a></li>
<li class="toctree-l1"><a class="reference internal" href="authors.html">Authors</a></li>
<li class="toctree-l1"><a class="reference internal" href="history.html">History</a></li>
<li class="toctree-l1"><a class="reference internal" href="license.html">obogaf::parser License</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">obogaf::parser</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>Tutorial</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/tutorial.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="tutorial">
<span id="id1"></span><h1>Tutorial<a class="headerlink" href="#tutorial" title="Permalink to this headline">¶</a></h1>
<p>Here we show a step-by-step application of <code class="docutils literal notranslate"><span class="pre">obogaf::parser</span></code> by using the Gene Ontology (<code class="docutils literal notranslate"><span class="pre">GO</span></code>) and the Human Phenotype Ontology (<code class="docutils literal notranslate"><span class="pre">HPO</span></code>) and their respective annotation file. The snippets of Perl code shown in the examples below are glued together respectively in the script <code class="docutils literal notranslate"><span class="pre">GOscript.pl</span></code> and <code class="docutils literal notranslate"><span class="pre">HPOscript.pl</span></code> stored in <code class="docutils literal notranslate"><span class="pre">script</span></code> folder of this repository.</p>
<hr class="docutils" />
<p>The experiments run on this tutorial were executed by using the <code class="docutils literal notranslate"><span class="pre">obogaf::parser</span></code> version <code class="docutils literal notranslate"><span class="pre">1.001</span></code>, the Perl version <code class="docutils literal notranslate"><span class="pre">5.22.1</span></code> and on a machine having Ubuntu 16.04 as operative system.</p>
<hr class="docutils" />
<div class="section" id="application-to-the-gene-ontology-go">
<h2>Application to the Gene Ontology (GO)<a class="headerlink" href="#application-to-the-gene-ontology-go" title="Permalink to this headline">¶</a></h2>
<p>For all the examples shown in this tutorial, we store I/O files in the directory <code class="docutils literal notranslate"><span class="pre">example/data</span></code>:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ mkdir -p example/data  <span class="c1">## create a directory if it does not already exist</span>
</pre></div>
</div>
<div class="section" id="parse-the-go-obo-file">
<h3>Parse the GO obo file<a class="headerlink" href="#parse-the-go-obo-file" title="Permalink to this headline">¶</a></h3>
<p>First of all we must download the <em>obo</em> file from the <a class="reference external" href="http://geneontology.org/docs/download-ontology/">Gene Ontoloy website</a>. We download the <em>basic</em> version of the <strong>GO</strong>, because this version excludes relationships that cross the 3 <strong>GO</strong> hierarchies (<code class="docutils literal notranslate"><span class="pre">BP</span></code>, <code class="docutils literal notranslate"><span class="pre">MF</span></code>, <code class="docutils literal notranslate"><span class="pre">CC</span></code>). To do that in a Linux environment, just type on the bash:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ <span class="nb">cd</span> example/data <span class="o">&amp;&amp;</span> wget http://purl.obolibrary.org/obo/go/go-basic.obo -O gobasic.obo
</pre></div>
</div>
<p>Let’s have a look to the <code class="docutils literal notranslate"><span class="pre">gobasic.obo</span></code> (release <code class="docutils literal notranslate"><span class="pre">01/07/19</span></code>) file to see how it is structured. For instance, to display the first <code class="docutils literal notranslate"><span class="pre">60</span></code> lines we can type on the Linux Shell <code class="docutils literal notranslate"><span class="pre">head</span> <span class="pre">-n60</span> <span class="pre">example/data/gobasic.obo</span></code>:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>format-version: 1.2
data-version: releases/2019-07-01
subsetdef: gocheck_do_not_annotate &quot;Term not to be used for direct annotation&quot;
subsetdef: gocheck_do_not_manually_annotate &quot;Term not to be used for direct manual annotation&quot;
subsetdef: goslim_agr &quot;AGR slim&quot;
subsetdef: goslim_aspergillus &quot;Aspergillus GO slim&quot;
subsetdef: goslim_candida &quot;Candida GO slim&quot;
subsetdef: goslim_chembl &quot;ChEMBL protein targets summary&quot;
subsetdef: goslim_flybase_ribbon &quot;FlyBase Drosophila GO ribbon slim&quot;
subsetdef: goslim_generic &quot;Generic GO slim&quot;
subsetdef: goslim_metagenomics &quot;Metagenomics GO slim&quot;
subsetdef: goslim_mouse &quot;Mouse GO slim&quot;
subsetdef: goslim_pir &quot;PIR GO slim&quot;
subsetdef: goslim_plant &quot;Plant GO slim&quot;
subsetdef: goslim_pombe &quot;Fission yeast GO slim&quot;
subsetdef: goslim_synapse &quot;synapse GO slim&quot;
subsetdef: goslim_yeast &quot;Yeast GO slim&quot;
synonymtypedef: syngo_official_label &quot;label approved by the SynGO project&quot;
synonymtypedef: systematic_synonym &quot;Systematic synonym&quot; EXACT
default-namespace: gene_ontology
remark: cvs version: use data-version
remark: Includes Ontology(OntologyID(OntologyIRI(&lt;http://purl.obolibrary.org/obo/go/never_in_taxon.owl&gt;))) [Axioms: 18 Logical Axioms: 0]
ontology: go

[Term]
id: GO:0000001
name: mitochondrion inheritance
namespace: biological_process
def: &quot;The distribution of mitochondria, including the mitochondrial genome, into daughter cells after mitosis or meiosis, mediated by interactions between mitochondria and the cytoskeleton.&quot; [GOC:mcc, PMID:10873824, PMID:11389764]
synonym: &quot;mitochondrial inheritance&quot; EXACT []
is_a: GO:0048308 ! organelle inheritance
is_a: GO:0048311 ! mitochondrion distribution

[Term]
id: GO:0000002
name: mitochondrial genome maintenance
namespace: biological_process
def: &quot;The maintenance of the structure and integrity of the mitochondrial genome; includes replication and segregation of the mitochondrial chromosome.&quot; [GOC:ai, GOC:vw]
is_a: GO:0007005 ! mitochondrion organization

[Term]
id: GO:0000003
name: reproduction
namespace: biological_process
alt_id: GO:0019952
alt_id: GO:0050876
def: &quot;The production of new individuals that contain some portion of genetic material inherited from one or more parent organisms.&quot; [GOC:go_curators, GOC:isa_complete, GOC:jl, ISBN:0198506732]
subset: goslim_agr
subset: goslim_chembl
subset: goslim_flybase_ribbon
subset: goslim_generic
subset: goslim_pir
subset: goslim_plant
synonym: &quot;reproductive physiological process&quot; EXACT []
xref: Wikipedia:Reproduction
is_a: GO:0008150 ! biological_process


[Term]
id: GO:0000005
name: obsolete ribosomal chaperone activity

... to be continued ...
</pre></div>
</div>
<p>To extrapolate the <strong>GO</strong> edges from the <code class="docutils literal notranslate"><span class="pre">gobasic.obo</span></code> file, we can use the subroutine <code class="docutils literal notranslate"><span class="pre">build_edges</span></code>. This subroutine receives in input the <code class="docutils literal notranslate"><span class="pre">obo</span></code> file:</p>
<div class="highlight-perl notranslate"><div class="highlight"><pre><span></span><span class="c1">## loading the obo file and calling the subroutine</span>
<span class="k">my</span> <span class="nv">$obofile</span><span class="o">=</span> <span class="s">&quot;example/data/gobasic.obo&quot;</span><span class="p">;</span>
<span class="k">my</span> <span class="nv">$gores</span><span class="o">=</span> <span class="nn">obogaf::parser::</span><span class="n">build_edges</span><span class="p">(</span><span class="nv">$obofile</span><span class="p">);</span>

<span class="c1">## storing</span>
<span class="k">my</span> <span class="nv">$goedges</span><span class="o">=</span> <span class="s">&quot;example/data/gobasic-edges.txt&quot;</span><span class="p">;</span>
<span class="nb">open</span> <span class="n">OUT</span><span class="p">,</span> <span class="s">&quot;&gt; $goedges&quot;</span><span class="p">;</span>
<span class="k">print</span> <span class="n">OUT</span> <span class="s">&quot;${$gores}&quot;</span><span class="p">;</span> <span class="c1">## dereferencing</span>
<span class="nb">close</span> <span class="n">OUT</span><span class="p">;</span>
</pre></div>
</div>
<p>For the sake of the space, below we just show the first <code class="docutils literal notranslate"><span class="pre">25</span></code> lines of the output file <code class="docutils literal notranslate"><span class="pre">gobasic-edges.txt</span></code> (<code class="docutils literal notranslate"><span class="pre">head</span> <span class="pre">-n25</span> <span class="pre">example/data/gobasic-edges.txt</span></code>):</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>biological_process  GO:0048308  GO:0000001  is-a
biological_process  GO:0048311  GO:0000001  is-a
biological_process  GO:0007005  GO:0000002  is-a
biological_process  GO:0008150  GO:0000003  is-a
molecular_function  GO:0005385  GO:0000006  is-a
molecular_function  GO:0005385  GO:0000007  is-a
molecular_function  GO:0000030  GO:0000009  is-a
molecular_function  GO:0016765  GO:0000010  is-a
biological_process  GO:0007033  GO:0000011  is-a
biological_process  GO:0048308  GO:0000011  is-a
biological_process  GO:0006281  GO:0000012  is-a
molecular_function  GO:0004520  GO:0000014  is-a
cellular_component  GO:0044445  GO:0000015  is-a
cellular_component  GO:1902494  GO:0000015  is-a
molecular_function  GO:0004553  GO:0000016  is-a
biological_process  GO:0042946  GO:0000017  is-a
biological_process  GO:0051052  GO:0000018  is-a
biological_process  GO:0000018  GO:0000019  is-a
biological_process  GO:0051231  GO:0000022  is-a
biological_process  GO:1903047  GO:0000022  is-a
biological_process  GO:0000070  GO:0000022  part-of
biological_process  GO:0007052  GO:0000022  part-of
biological_process  GO:0005984  GO:0000023  is-a
biological_process  GO:0000023  GO:0000024  is-a
biological_process  GO:0046351  GO:0000024  is-a

... to be continued ...
</pre></div>
</div>
<p>The first column of the output file refers to the domain whose a <strong>GO</strong> term belong to, the second and the third column represent the edge as pair of nodes in the form <code class="docutils literal notranslate"><span class="pre">source</span> <span class="pre">(parent)</span> <span class="pre">-</span> <span class="pre">destination</span> <span class="pre">(child)</span></code> and the fourth column refers to the kind of relationships. This column can assume only two values, <code class="docutils literal notranslate"><span class="pre">is-a</span></code> and <code class="docutils literal notranslate"><span class="pre">part-of</span></code>, since it is safe grouping annotations by using both these relationships. For more details about <strong>GO</strong> relationships have a look at this <a class="reference external" href="http://geneontology.org/docs/ontology-relations/">link</a>.</p>
<p>To isolate nodes and relationships belonging to one of the <strong>GO</strong> sub-ontology (e.g. <code class="docutils literal notranslate"><span class="pre">biological_process</span> <span class="pre">(BP)</span></code>), we can use the subroutine <code class="docutils literal notranslate"><span class="pre">build_subonto</span></code>. This subroutine receives in input the edges file obtained by calling <code class="docutils literal notranslate"><span class="pre">build_edges</span></code> and the specific sub-domain for which we want to extrapolate edges.</p>
<div class="highlight-perl notranslate"><div class="highlight"><pre><span></span><span class="c1">## loading and calling</span>
<span class="k">my</span> <span class="nv">$goedges</span><span class="o">=</span> <span class="s">&quot;example/data/gobasic-edges.txt&quot;</span><span class="p">;</span> <span class="c1">## obtained previously by calling obogaf::parser::build_edges</span>
<span class="k">my</span> <span class="nv">$BPres</span><span class="o">=</span> <span class="nn">obogaf::parser::</span><span class="n">build_subonto</span><span class="p">(</span><span class="nv">$goedges</span><span class="p">,</span> <span class="s">&quot;biological_process&quot;</span><span class="p">);</span>

<span class="c1">## storing</span>
<span class="k">my</span> <span class="nv">$BPedges</span><span class="o">=</span> <span class="s">&quot;example/data/gobasic-edgesBP.txt&quot;</span><span class="p">;</span>
<span class="nb">open</span> <span class="n">OUT</span><span class="p">,</span> <span class="s">&quot;&gt; $BPedges&quot;</span><span class="p">;</span>
<span class="k">print</span> <span class="n">OUT</span> <span class="s">&quot;${$BPres}&quot;</span><span class="p">;</span>
<span class="nb">close</span> <span class="n">OUT</span><span class="p">;</span>
</pre></div>
</div>
<p>Below we report the first <code class="docutils literal notranslate"><span class="pre">10</span></code> lines of <code class="docutils literal notranslate"><span class="pre">gobasic-edgesBP.txt</span></code> (<code class="docutils literal notranslate"><span class="pre">head</span> <span class="pre">-n10</span> <span class="pre">example/data/gobasic-edgesBP.txt</span></code>):</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>GO:0048308  GO:0000001  is-a
GO:0048311  GO:0000001  is-a
GO:0007005  GO:0000002  is-a
GO:0008150  GO:0000003  is-a
GO:0007033  GO:0000011  is-a
GO:0048308  GO:0000011  is-a
GO:0006281  GO:0000012  is-a
GO:0042946  GO:0000017  is-a
GO:0051052  GO:0000018  is-a
GO:0000018  GO:0000019  is-a

... to be continued ...
</pre></div>
</div>
<p>It is worth noting that the same output can be also achieved by using the <code class="docutils literal notranslate"><span class="pre">grep</span></code> command (in a Linux environment):</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ grep <span class="s2">&quot;biological_process&quot;</span> example/data/gobasic-edges.txt <span class="p">|</span> cut -f2- &gt; example/data/gobasic-edgesBP.txt
</pre></div>
</div>
<p>If we want to isolate nodes and relationships separately for each <strong>GO</strong> subontology at one fell swoop, by Perl:</p>
<div class="highlight-perl notranslate"><div class="highlight"><pre><span></span><span class="k">my</span> <span class="nv">$goedges</span><span class="o">=</span> <span class="s">&quot;example/data/gobasic-edges.txt&quot;</span><span class="p">;</span> <span class="c1">## obtained previously by calling obogaf::parser::build_edges</span>
<span class="k">my</span> <span class="nv">@domains</span><span class="o">=</span> <span class="sx">qw(biological_process molecular_function cellular_component)</span><span class="p">;</span>
<span class="k">my</span> <span class="nv">%aspects</span><span class="o">=</span><span class="p">(</span><span class="n">biological_process</span> <span class="o">=&gt;</span> <span class="s">&quot;BP&quot;</span><span class="p">,</span> <span class="n">molecular_function</span> <span class="o">=&gt;</span> <span class="s">&quot;MF&quot;</span><span class="p">,</span> <span class="n">cellular_component</span> <span class="o">=&gt;</span> <span class="s">&quot;CC&quot;</span><span class="p">);</span>

<span class="k">foreach</span> <span class="k">my</span> <span class="nv">$domain</span> <span class="p">(</span><span class="nv">@domains</span><span class="p">){</span>
    <span class="k">my</span> <span class="nv">$outfile</span><span class="o">=</span> <span class="s">&quot;example/data/gobasic-edges&quot;</span><span class="o">.</span><span class="s">&quot;$aspects{$domain}&quot;</span><span class="o">.</span><span class="s">&quot;.txt&quot;</span><span class="p">;</span>
    <span class="nb">open</span> <span class="n">OUT</span><span class="p">,</span> <span class="s">&quot;&gt; $outfile&quot;</span><span class="p">;</span>
    <span class="k">my</span> <span class="nv">$domainres</span><span class="o">=</span> <span class="nn">obogaf::parser::</span><span class="n">build_subonto</span><span class="p">(</span><span class="nv">$goedges</span><span class="p">,</span> <span class="nv">$domain</span><span class="p">);</span>
    <span class="k">print</span> <span class="n">OUT</span> <span class="s">&quot;${$domainres}&quot;</span><span class="p">;</span>
    <span class="nb">close</span> <span class="n">OUT</span><span class="p">;</span>
<span class="p">}</span>
</pre></div>
</div>
<p>and by bash:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nv">goedges</span><span class="o">=</span><span class="s2">&quot;example/data/gobasic-edges.txt&quot;</span><span class="p">;</span> <span class="c1">## obtained previously by calling obogaf::parser::build_edges</span>
<span class="nv">domains</span><span class="o">=(</span><span class="s2">&quot;biological_process&quot;</span> <span class="s2">&quot;molecular_function&quot;</span> <span class="s2">&quot;cellular_component&quot;</span><span class="o">)</span><span class="p">;</span>
<span class="nv">aspects</span><span class="o">=(</span><span class="s2">&quot;BP&quot;</span> <span class="s2">&quot;MF&quot;</span> <span class="s2">&quot;CC&quot;</span><span class="o">)</span><span class="p">;</span>

<span class="nv">len</span><span class="o">=</span><span class="s2">&quot;</span><span class="si">${#</span><span class="nv">domains</span><span class="p">[@]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">;</span>
<span class="k">for</span> <span class="o">((</span><span class="nv">i</span> <span class="o">=</span> <span class="m">0</span> <span class="p">;</span> i &lt; len <span class="p">;</span> i++<span class="o">))</span><span class="p">;</span> <span class="k">do</span>
    grep <span class="si">${</span><span class="nv">domains</span><span class="p">[</span><span class="nv">$i</span><span class="p">]</span><span class="si">}</span> example/data/gobasic-edges.txt <span class="p">|</span> cut -f2- &gt; example/data/gobasic-edges<span class="si">${</span><span class="nv">aspects</span><span class="p">[</span><span class="nv">$i</span><span class="p">]</span><span class="si">}</span>.txt
<span class="k">done</span>
</pre></div>
</div>
<p>To print some statistics on the <code class="docutils literal notranslate"><span class="pre">GO</span></code> graph, we can use the subroutine <code class="docutils literal notranslate"><span class="pre">make_stat</span></code>. The input arguments required by this subroutine are:</p>
<ol class="arabic simple">
<li><code class="docutils literal notranslate"><span class="pre">$goedges</span></code>: file containing the <code class="docutils literal notranslate"><span class="pre">GO</span></code> graph represented as a list of edges where each edge is turn represented as a pair of vertices <code class="docutils literal notranslate"><span class="pre">tab</span></code> separated (<code class="docutils literal notranslate"><span class="pre">$goedges</span></code> file can be obtained by calling the <code class="docutils literal notranslate"><span class="pre">build_edges</span></code> subroutine)</li>
<li><code class="docutils literal notranslate"><span class="pre">$parentIndex</span></code> and <code class="docutils literal notranslate"><span class="pre">$childIndex</span></code>: index referring restrictively to the column containing the <code class="docutils literal notranslate"><span class="pre">source</span></code> and <code class="docutils literal notranslate"><span class="pre">destination</span></code> nodes in the <code class="docutils literal notranslate"><span class="pre">$goedges</span></code> file (reminder: Perl starts counting from zero).</li>
</ol>
<div class="highlight-perl notranslate"><div class="highlight"><pre><span></span><span class="k">my</span> <span class="p">(</span><span class="nv">$goedges</span><span class="p">,</span> <span class="nv">$parentIndex</span><span class="p">,</span> <span class="nv">$childIndex</span><span class="p">)</span><span class="o">=</span> <span class="p">(</span><span class="s">&quot;example/data/gobasic-edges.txt&quot;</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">);</span>
<span class="k">my</span> <span class="nv">$res</span><span class="o">=</span> <span class="nn">obogaf::parser::</span><span class="n">make_stat</span><span class="p">(</span><span class="nv">$goedges</span><span class="p">,</span> <span class="nv">$parentIndex</span><span class="p">,</span> <span class="nv">$childIndex</span><span class="p">);</span>
<span class="k">print</span> <span class="s">&quot;$res&quot;</span><span class="p">;</span>

<span class="c1">## results printed on the shell</span>
<span class="n">oboterm</span> <span class="sr">&lt;tab&gt;</span> <span class="n">degree</span> <span class="sr">&lt;tab&gt;</span> <span class="n">indegree</span> <span class="sr">&lt;tab&gt;</span> <span class="n">outdegree</span>
<span class="n">GO:0032991</span>  <span class="mi">470</span> <span class="mi">1</span>   <span class="mi">469</span>
<span class="n">GO:0016616</span>  <span class="mi">351</span> <span class="mi">1</span>   <span class="mi">350</span>
<span class="n">GO:0016709</span>  <span class="mi">303</span> <span class="mi">2</span>   <span class="mi">301</span>
<span class="n">GO:0044444</span>  <span class="mi">211</span> <span class="mi">2</span>   <span class="mi">209</span>
<span class="n">GO:0016758</span>  <span class="mi">204</span> <span class="mi">1</span>   <span class="mi">203</span>
<span class="n">GO:0048856</span>  <span class="mi">199</span> <span class="mi">1</span>   <span class="mi">198</span>
<span class="n">GO:0098797</span>  <span class="mi">180</span> <span class="mi">2</span>   <span class="mi">178</span>
<span class="n">GO:0003006</span>  <span class="mi">171</span> <span class="mi">2</span>   <span class="mi">169</span>
<span class="n">GO:0016747</span>  <span class="mi">159</span> <span class="mi">1</span>   <span class="mi">158</span>
<span class="o">.</span>
<span class="o">.</span>
<span class="o">.</span>
<span class="o">~</span><span class="n">summary</span> <span class="n">stats</span><span class="o">~</span>
<span class="n">nodes:</span> <span class="mi">44945</span>
<span class="n">edges:</span> <span class="mi">83839</span>
<span class="n">max</span> <span class="n">degree:</span> <span class="mi">470</span>
<span class="n">min</span> <span class="n">degree:</span> <span class="mi">1</span>
<span class="n">median</span> <span class="n">degree:</span> <span class="mf">2.0000</span>
<span class="n">average</span> <span class="n">degree:</span> <span class="mf">1.8654</span>
<span class="n">density:</span> <span class="mf">4.1504e-05</span>
</pre></div>
</div>
<p>As we can observe from the snippet above, for each node of the graph, <code class="docutils literal notranslate"><span class="pre">degree</span></code>, <code class="docutils literal notranslate"><span class="pre">in-degree</span></code> and <code class="docutils literal notranslate"><span class="pre">out-degree</span></code> are printed. Nodes are sorted in a decreasing order on the basis of degree, from the higher to the smaller one. In addition the following statistics are also returned: 1) number of nodes and edges of the graph; 2) maximum and minimum degree; 3) average and median degree; 4) density of the graph.</p>
<p>To compute the stats just for a specific <code class="docutils literal notranslate"><span class="pre">GO</span></code> subontology (e.g. <code class="docutils literal notranslate"><span class="pre">GO</span> <span class="pre">BP</span></code>) we can always use <code class="docutils literal notranslate"><span class="pre">make_stat</span></code>, by properly setting its input arguments:</p>
<div class="highlight-perl notranslate"><div class="highlight"><pre><span></span><span class="k">my</span> <span class="p">(</span><span class="nv">$goedges</span><span class="p">,</span> <span class="nv">$parentIndex</span><span class="p">,</span> <span class="nv">$childIndex</span><span class="p">)</span><span class="o">=</span> <span class="p">(</span><span class="s">&quot;example/data/gobasic-edgesBP.txt&quot;</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">);</span>
<span class="k">my</span> <span class="nv">$res</span><span class="o">=</span> <span class="nn">obogaf::parser::</span><span class="n">make_stat</span><span class="p">(</span><span class="nv">$goedges</span><span class="p">,</span> <span class="nv">$parentIndex</span><span class="p">,</span> <span class="nv">$childIndex</span><span class="p">);</span>
<span class="k">print</span> <span class="s">&quot;$res&quot;</span><span class="p">;</span>

<span class="c1">## results returned on the shell</span>
<span class="n">oboterm</span> <span class="sr">&lt;tab&gt;</span> <span class="n">degree</span> <span class="sr">&lt;tab&gt;</span> <span class="n">indegree</span> <span class="sr">&lt;tab&gt;</span> <span class="n">outdegree</span>
<span class="n">GO:0048856</span>  <span class="mi">199</span> <span class="mi">1</span>   <span class="mi">198</span>
<span class="n">GO:0003006</span>  <span class="mi">171</span> <span class="mi">2</span>   <span class="mi">169</span>
<span class="n">GO:0051241</span>  <span class="mi">136</span> <span class="mi">2</span>   <span class="mi">134</span>
<span class="n">GO:0051240</span>  <span class="mi">129</span> <span class="mi">2</span>   <span class="mi">127</span>
<span class="n">GO:0014070</span>  <span class="mi">128</span> <span class="mi">1</span>   <span class="mi">127</span>
<span class="n">GO:1901700</span>  <span class="mi">112</span> <span class="mi">1</span>   <span class="mi">111</span>
<span class="n">GO:0022414</span>  <span class="mi">110</span> <span class="mi">2</span>   <span class="mi">108</span>
<span class="n">GO:0048646</span>  <span class="mi">108</span> <span class="mi">2</span>   <span class="mi">106</span>
<span class="n">GO:0031328</span>  <span class="mi">105</span> <span class="mi">3</span>   <span class="mi">102</span>
<span class="n">GO:1901361</span>  <span class="mi">105</span> <span class="mi">2</span>   <span class="mi">103</span>
<span class="o">.</span>
<span class="o">.</span>
<span class="o">.</span>
<span class="o">~</span><span class="n">summary</span> <span class="n">stats</span><span class="o">~</span>
<span class="n">nodes:</span> <span class="mi">29652</span>
<span class="n">edges:</span> <span class="mi">62739</span>
<span class="n">max</span> <span class="n">degree:</span> <span class="mi">199</span>
<span class="n">min</span> <span class="n">degree:</span> <span class="mi">1</span>
<span class="n">median</span> <span class="n">degree:</span> <span class="mf">3.0000</span>
<span class="n">average</span> <span class="n">degree:</span> <span class="mf">2.1158</span>
<span class="n">density:</span> <span class="mf">7.1358e-05</span>
</pre></div>
</div>
</div>
<div class="section" id="parse-the-goa-annotation-file">
<h3>Parse the GOA annotation file<a class="headerlink" href="#parse-the-goa-annotation-file" title="Permalink to this headline">¶</a></h3>
<p><code class="docutils literal notranslate"><span class="pre">obogaf::parser</span></code> can be also used to parse the annotation file taken from the Gene Ontology Annotation (<code class="docutils literal notranslate"><span class="pre">GOA</span></code>) Database (<a class="reference external" href="https://www.ebi.ac.uk/GOA">link</a>).</p>
<p>For the examples shown below we use the annotation file of the <code class="docutils literal notranslate"><span class="pre">CHICKEN</span></code> model organism (release <code class="docutils literal notranslate"><span class="pre">7/29/19</span></code>), but of course <code class="docutils literal notranslate"><span class="pre">obogaf::parser</span></code> subroutines can be applied to parse the annotation file of any other organisms listed in the <code class="docutils literal notranslate"><span class="pre">GOA</span></code> database and more in general to parse any file structured as those listed in the <code class="docutils literal notranslate"><span class="pre">GOA</span></code> database.</p>
<hr class="docutils" />
<p>NOTE: the annotation file on <code class="docutils literal notranslate"><span class="pre">GOA</span></code> website are monthly updated. The release used at the time of writing this tutorial is July release (7/29/19).</p>
<hr class="docutils" />
<p>First we must download the annotation file in the <code class="docutils literal notranslate"><span class="pre">example/data</span></code> folder (note that the link show below refers to the most updated release):</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ <span class="nb">cd</span> example/data <span class="o">&amp;&amp;</span> wget ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/CHICKEN/goa_chicken.gaf.gz -O goa_chicken.gaf.gz
</pre></div>
</div>
<p>By having a look to the <code class="docutils literal notranslate"><span class="pre">goa_chicken.gaf.gz</span></code> file we see that it is structured as follow (for the sake of space we display just the first <code class="docutils literal notranslate"><span class="pre">20</span></code> lines):</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>!gaf-version: 2.1
!
!The set of protein accessions included in this file is based on UniProt reference proteomes, which provide one protein per gene.
!They include the protein sequences annotated in Swiss-Prot or the longest TrEMBL transcript if there is no Swiss-Prot record.
!If a particular protein accession is not annotated with GO, then it will not appear in this file.
!
!Note that the annotation set in this file is filtered in order to reduce redundancy; the full, unfiltered set can be found in
!ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/goa_uniprot_all.gz
!
!Generated: 2019-07-29 13:51
!GO-version: http://purl.obolibrary.org/obo/go/releases/2019-07-17/extensions/go-plus.owl
!
UniProtKB   A0A088BIK7  EDbeta      GO:0005200  GO_REF:0000002  IEA InterPro:IPR003461  F   Keratin EDbeta|EDBETA   protein taxon:9031  20190727    InterPro
UniProtKB   A0A088BIK7  EDbeta      GO:0005882  GO_REF:0000038  IEA UniProtKB-KW:KW-0416    C   Keratin EDbeta|EDBETA   protein taxon:9031  20190727    UniProt
UniProtKB   A0A088BIK7  EDbeta      GO:0007010  GO_REF:0000108  IEA GO:0005200  P   Keratin EDbeta|EDBETA   protein taxon:9031  20190727    GOC
UniProtKB   A0A0A0MQ32  LOXL2       GO:0000122  GO_REF:0000107  IEA UniProtKB:Q9Y4K0|ensembl:ENSP00000373783    P   Lysyl oxidase homolog 2 LOXL2   protein taxon:9031  20190727    Ensembl
UniProtKB   A0A0A0MQ32  LOXL2       GO:0000785  GO_REF:0000107  IEA UniProtKB:Q9Y4K0|ensembl:ENSP00000373783    C   Lysyl oxidase homolog 2 LOXL2   protein taxon:9031  20190727    Ensembl
UniProtKB   A0A0A0MQ32  LOXL2       GO:0001666  GO_REF:0000107  IEA UniProtKB:P58022|ensembl:ENSMUSP00000022660 P   Lysyl oxidase homolog 2 LOXL2   protein taxon:9031  20190727    Ensembl
UniProtKB   A0A0A0MQ32  LOXL2       GO:0001837  GO_REF:0000107  IEA UniProtKB:Q9Y4K0|ensembl:ENSP00000373783    P   Lysyl oxidase homolog 2 LOXL2   protein taxon:9031  20190727    Ensembl
UniProtKB   A0A0A0MQ32  LOXL2       GO:0001935  GO_REF:0000107  IEA UniProtKB:Q9Y4K0|ensembl:ENSP00000373783    P   Lysyl oxidase homolog 2 LOXL2   protein taxon:9031  20190727    Ensembl

... to be continued ...
</pre></div>
</div>
<p>Now we can build the list of annotations by using the subroutine <code class="docutils literal notranslate"><span class="pre">gene2biofun</span></code>. The input arguments required are:</p>
<ol class="arabic simple">
<li><code class="docutils literal notranslate"><span class="pre">$inputfile</span></code>: <code class="docutils literal notranslate"><span class="pre">GOA</span></code> annotation file for the <code class="docutils literal notranslate"><span class="pre">CHICKEN</span></code> organism;</li>
<li><code class="docutils literal notranslate"><span class="pre">$geneindex</span></code>: and <code class="docutils literal notranslate"><span class="pre">$geneindex</span></code>: index referring respectively to the column containing the proteins and the <code class="docutils literal notranslate"><span class="pre">GO</span></code> term in the <code class="docutils literal notranslate"><span class="pre">$inputfile</span></code> file.</li>
</ol>
<div class="highlight-perl notranslate"><div class="highlight"><pre><span></span><span class="k">my</span> <span class="p">(</span><span class="nv">$inputfile</span><span class="p">,</span> <span class="nv">$geneindex</span><span class="p">,</span> <span class="nv">$classindex</span><span class="p">)</span><span class="o">=</span> <span class="p">(</span><span class="s">&quot;example/data/goa_chicken.gaf.gz&quot;</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">);</span>
<span class="k">my</span> <span class="p">(</span><span class="nv">$res</span><span class="p">,</span> <span class="nv">$stat</span><span class="p">)</span><span class="o">=</span> <span class="nn">obogaf::parser::</span><span class="n">gene2biofun</span><span class="p">(</span><span class="nv">$inputfile</span><span class="p">,</span> <span class="nv">$geneindex</span><span class="p">,</span> <span class="nv">$classindex</span><span class="p">);</span>

<span class="k">my</span> <span class="nv">$goaout</span><span class="o">=</span> <span class="s">&quot;example/data/chicken.uniprot2go.txt&quot;</span><span class="p">;</span>
<span class="nb">open</span> <span class="n">OUT</span><span class="p">,</span> <span class="s">&quot;&gt; $goaout&quot;</span><span class="p">;</span>
<span class="k">foreach</span> <span class="k">my</span> <span class="nv">$k</span> <span class="p">(</span><span class="nb">sort</span><span class="p">{</span><span class="nv">$a</span> <span class="ow">cmp</span> <span class="nv">$b</span><span class="p">}</span> <span class="nb">keys</span> <span class="nv">%$res</span><span class="p">)</span> <span class="p">{</span> <span class="k">print</span> <span class="n">OUT</span> <span class="s">&quot;$k $$res{$k}\n&quot;</span><span class="p">;}</span>
<span class="nb">close</span> <span class="n">OUT</span><span class="p">;</span>
<span class="k">print</span> <span class="s">&quot;${$stat}\n&quot;</span><span class="p">;</span>

<span class="c1">## results printed on the shell</span>
<span class="n">genes:</span> <span class="mi">15695</span>
<span class="n">ontology</span> <span class="n">terms:</span> <span class="mi">13953</span>
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">gene2biofun</span></code> returns a list of two anonymous references. The first is an anonymous hash storing for each UniProtKB protein all its associated <code class="docutils literal notranslate"><span class="pre">GO</span></code> terms (pipe separated). The second is an anonymous scalar containing basic statistics such as the total unique number of proteins and ontology terms. In the example above the anonymous hash is addressed in the output file <code class="docutils literal notranslate"><span class="pre">example/data/chicken.uniprot2go.txt</span></code> and the stats are printed on the shell. Finally, it is worth noting that <code class="docutils literal notranslate"><span class="pre">gene2biofun</span></code> can handle both compress <code class="docutils literal notranslate"><span class="pre">.gz</span></code> file and plain <code class="docutils literal notranslate"><span class="pre">.txt</span></code> file. Below we report as an example a snapshot of the associations between UniProtKB entry and <code class="docutils literal notranslate"><span class="pre">GO</span></code> terms obtained by running <code class="docutils literal notranslate"><span class="pre">gene2biofun</span></code> and stored in the file <code class="docutils literal notranslate"><span class="pre">example/data/chicken.uniprot2go.txt</span></code> (<code class="docutils literal notranslate"><span class="pre">head</span> <span class="pre">-n10</span> <span class="pre">example/data/chicken.uniprot2go.txt</span></code>):</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>A0A088BIK7 GO:0005200|GO:0005882|GO:0007010
A0A0A0MQ32 GO:0000122|GO:0000785|GO:0001666|GO:0001837|GO:0001935|GO:0002040|GO:0004720|GO:0005044|GO:0005507|GO:0005509|GO:0005615|GO:0005654|GO:0005783|GO:0006897|GO:0010718|GO:0016020|GO:0018057|GO:0030199|GO:0032332|GO:0043542|GO:0046688|GO:0070492|GO:0070828|GO:1902455
A0A0A0MQ34 GO:0009374
A0A0A0MQ35 GO:0000421|GO:0005654|GO:0005765|GO:0016021|GO:0032266|GO:0097352
A0A0A0MQ36 GO:0005246|GO:0005509|GO:0007165
A0A0A0MQ42 GO:0005654|GO:0005794|GO:0019221|GO:0030368
A0A0A0MQ45 GO:0000086|GO:0004674|GO:0005524|GO:0005634|GO:0005654|GO:0005813|GO:0007147|GO:0018105|GO:0032154|GO:0032515|GO:0035556|GO:0051726|GO:1904668
A0A0A0MQ47 GO:0000122|GO:0000993|GO:0002039|GO:0005634|GO:0005829|GO:0008285|GO:0010452|GO:0018026|GO:0018027|GO:0043516|GO:0046975
A0A0A0MQ52 GO:0000724|GO:0003678|GO:0003682|GO:0003682|GO:0003688|GO:0003697|GO:0005524|GO:0005634|GO:0006270|GO:0007292|GO:0019899|GO:0032406|GO:0032407|GO:0032408|GO:0032508|GO:0036298|GO:0036298|GO:0042555|GO:0070716|GO:0070716|GO:0071168|GO:0097362|GO:0097362
A0A0A0MQ56 GO:0005615|GO:0005623|GO:0010975|GO:1990830|GO:0005874

... to be continued...
</pre></div>
</div>
</div>
<div class="section" id="map-go-terms-between-releases">
<h3>Map GO terms between releases<a class="headerlink" href="#map-go-terms-between-releases" title="Permalink to this headline">¶</a></h3>
<p>In time-lapse hold-out experiments we use annotations of an old <code class="docutils literal notranslate"><span class="pre">GO</span></code> release to predict the protein function of a more recent <code class="docutils literal notranslate"><span class="pre">GO</span></code> release. However, between different <code class="docutils literal notranslate"><span class="pre">GO</span></code> releases some ontology terms could be removed, others changed or become obsolete. Then before beginning time-lapse hold-out experiments, we need to map the old <code class="docutils literal notranslate"><span class="pre">GO</span></code> terms to the new ones by parsing the annotation file of an <em>old</em> <code class="docutils literal notranslate"><span class="pre">GO</span></code> release using as <strong>key</strong> the <em>alt-ID</em> taken from the obo file of the <em>new</em> <code class="docutils literal notranslate"><span class="pre">GO</span></code> release . The subroutine <code class="docutils literal notranslate"><span class="pre">map_OBOterm_between_release</span></code> does that for us.</p>
<p>Firstly, we must download the old annotation file of the <code class="docutils literal notranslate"><span class="pre">CHICKEN</span></code> organism in the <code class="docutils literal notranslate"><span class="pre">example/data</span></code> directory (here we use the <code class="docutils literal notranslate"><span class="pre">07/06/16</span></code> release):</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ <span class="nb">cd</span> example/data <span class="o">&amp;&amp;</span> wget ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/old/CHICKEN/goa_chicken.gaf.128.gz -O goa_chicken.gaf.128.gz
</pre></div>
</div>
<p>The input arguments required by <code class="docutils literal notranslate"><span class="pre">map_OBOterm_between_release</span></code> are:</p>
<ol class="arabic simple">
<li><code class="docutils literal notranslate"><span class="pre">$obofile</span></code>: the <em>new</em> release of a <code class="docutils literal notranslate"><span class="pre">GO</span></code> obo file (here we use the <code class="docutils literal notranslate"><span class="pre">01/07/19</span></code> release). This file is used to make the <code class="docutils literal notranslate"><span class="pre">alt_id</span> <span class="pre">-</span> <span class="pre">id</span></code> pairing by using <code class="docutils literal notranslate"><span class="pre">alt_id</span></code> as key;</li>
<li><code class="docutils literal notranslate"><span class="pre">$goafileOld</span></code>: the <em>old</em> release of an annotation file (for this example we use <code class="docutils literal notranslate"><span class="pre">07/06/16</span></code> release);</li>
<li><code class="docutils literal notranslate"><span class="pre">$classindex</span></code>: the index referring to the column of the <code class="docutils literal notranslate"><span class="pre">$goafileOld</span></code> containing the ontology terms to be mapped (in the <code class="docutils literal notranslate"><span class="pre">GOA</span></code> file the <code class="docutils literal notranslate"><span class="pre">GO</span></code> terms are in the 4 columns – NB: we must start to count from zero).</li>
</ol>
<div class="highlight-perl notranslate"><div class="highlight"><pre><span></span><span class="k">my</span> <span class="p">(</span><span class="nv">$obofile</span><span class="p">,</span> <span class="nv">$goafileOld</span><span class="p">,</span> <span class="nv">$classindex</span><span class="p">)</span><span class="o">=</span> <span class="p">(</span><span class="s">&quot;example/data/gobasic.obo&quot;</span><span class="p">,</span> <span class="s">&quot;example/data/goa_chicken.gaf.128.gz&quot;</span><span class="p">,</span> <span class="mi">4</span><span class="p">);</span>
<span class="k">my</span> <span class="p">(</span><span class="nv">$res</span><span class="p">,</span> <span class="nv">$stat</span><span class="p">)</span><span class="o">=</span> <span class="nn">obogaf::parser::</span><span class="n">map_OBOterm_between_release</span><span class="p">(</span><span class="nv">$obofile</span><span class="p">,</span> <span class="nv">$goafileOld</span><span class="p">,</span> <span class="nv">$classindex</span><span class="p">);</span>

<span class="k">my</span> <span class="nv">$mapfile</span><span class="o">=</span> <span class="s">&quot;example/data/chicken.goa.mapped.txt&quot;</span><span class="p">;</span>
<span class="nb">open</span> <span class="n">OUT</span><span class="p">,</span> <span class="s">&quot;&gt; $mapfile&quot;</span><span class="p">;</span>
<span class="k">print</span> <span class="n">OUT</span> <span class="s">&quot;${$res}&quot;</span><span class="p">;</span>
<span class="nb">close</span> <span class="n">OUT</span><span class="p">;</span>
<span class="k">print</span> <span class="s">&quot;${$stat}&quot;</span><span class="p">;</span>

<span class="c1"># results printed on the shell</span>
<span class="c1">#alt-id &lt;tab&gt; id</span>
<span class="n">GO:0000042</span>  <span class="n">GO:0034067</span>
<span class="n">GO:0000975</span>  <span class="n">GO:0044212</span>
<span class="n">GO:0000982</span>  <span class="n">GO:0000981</span>
<span class="n">GO:0000983</span>  <span class="n">GO:0016251</span>
<span class="n">GO:0001075</span>  <span class="n">GO:0016251</span>
<span class="n">GO:0001077</span>  <span class="n">GO:0001228</span>
<span class="n">GO:0001078</span>  <span class="n">GO:0001227</span>
<span class="n">GO:0001104</span>  <span class="n">GO:0003712</span>
<span class="n">GO:0001105</span>  <span class="n">GO:0003713</span>
<span class="n">GO:0001106</span>  <span class="n">GO:0003714</span>
<span class="o">.</span>
<span class="o">.</span>
<span class="o">.</span>
<span class="n">Tot</span><span class="o">.</span> <span class="n">ontology</span> <span class="n">terms:</span>    <span class="mi">12546</span>
<span class="n">Tot</span><span class="o">.</span> <span class="n">altID:</span> <span class="mi">2532</span>
<span class="n">Tot</span><span class="o">.</span> <span class="n">altID</span> <span class="n">seen:</span>    <span class="mi">201</span>
<span class="n">Tot</span><span class="o">.</span> <span class="n">altID</span> <span class="n">unseen:</span>  <span class="mi">2331</span>
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">map_OBOterm_between_release</span></code> subroutine returns a list of two anonymous references. The first is an anonymous scalar storing the annotations file in the same format of the input file but with the <em>obsolete</em> ontology terms substituted with the <em>updated</em> ones. The second reference is an anonymous scalar containing some basic statistics, such as the total unique number of ontology terms (of the old release) and the total number of mapped and unmapped <em>altID</em> ontology terms. In addition, all the found pairs <code class="docutils literal notranslate"><span class="pre">alt_id</span> <span class="pre">-</span> <span class="pre">id</span></code> are returned. In the example run above the anonymous hash is addressed in the output file <code class="docutils literal notranslate"><span class="pre">example/data/chicken.goa.mapped.txt</span></code> whereas the stats are printed on the shell.</p>
<p>The difference between the <em>old</em> and the <em>mapped</em> file can be easily displayed by using the <code class="docutils literal notranslate"><span class="pre">diff</span></code> command (in a Linux environment):</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ <span class="nb">cd</span> example/data <span class="o">&amp;&amp;</span> gunzip -k goa_chicken.gaf.128.gz
$ diff goa_chicken.gaf.128 chicken.goa.mapped.txt &gt; ann.diff
</pre></div>
</div>
<p>To give an example, below we show the first <code class="docutils literal notranslate"><span class="pre">23</span></code> lines of the file <code class="docutils literal notranslate"><span class="pre">ann.diff</span></code>:</p>
<div class="highlight-diff notranslate"><div class="highlight"><pre><span></span>75c75
&lt; UniProtKB A0AVX7  TESC        GO:0072661  GO_REF:0000024  ISS UniProtKB:Q96BS2    P   Calcineurin B homologous protein 3  CHP3_CHICK|TESC|CHP3    protein taxon:9031  20120627    UniProt
<span class="gd">---</span>
&gt; UniProtKB A0AVX7  TESC        GO:0072659  GO_REF:0000024  ISS UniProtKB:Q96BS2    P   Calcineurin B homologous protein 3  CHP3_CHICK|TESC|CHP3    protein taxon:9031  20120627    UniProt
159c159
&lt; UniProtKB A1DYI3  Wnt3        GO:0005578  GO_REF:0000040  IEA UniProtKB-SubCell:SL-0111   C   Protein Wnt A1DYI3_CHICK|Wnt3|WNT3  protein taxon:9031  20160507    UniProt
<span class="gd">---</span>
&gt; UniProtKB A1DYI3  Wnt3        GO:0031012  GO_REF:0000040  IEA UniProtKB-SubCell:SL-0111   C   Protein Wnt A1DYI3_CHICK|Wnt3|WNT3  protein taxon:9031  20160507    UniProt
234,235c234,235
&lt; UniProtKB A1KXM5  SPERT       GO:0016023  GO_REF:0000019  IEA Ensembl:ENSMUSP00000127439  C   Spermatid-associated protein    SPERT_CHICK|SPERT   protein taxon:9031  20160507    Ensembl
&lt; UniProtKB A1XGV6  TNFRSF19        GO:0004872  GO_REF:0000033  IBA PANTHER:PTN000950406    F   Troy-long   A1XGV6_CHICK|TNFRSF19   protein taxon:9031  20160114    GO_Central
<span class="gd">---</span>
&gt; UniProtKB A1KXM5  SPERT       GO:0031410  GO_REF:0000019  IEA Ensembl:ENSMUSP00000127439  C   Spermatid-associated protein    SPERT_CHICK|SPERT   protein taxon:9031  20160507    Ensembl
&gt; UniProtKB A1XGV6  TNFRSF19        GO:0038023  GO_REF:0000033  IBA PANTHER:PTN000950406    F   Troy-long   A1XGV6_CHICK|TNFRSF19   protein taxon:9031  20160114    GO_Central
268c268
&lt; UniProtKB A3F962  MBNL2       GO:0044822  GO_REF:0000019  IEA Ensembl:ENSP00000365861 F   Muscleblind-like 2 isoform 1    A3F962_CHICK|MBNL2  protein taxon:9031  20160507    Ensembl
<span class="gd">---</span>
&gt; UniProtKB A3F962  MBNL2       GO:0003723  GO_REF:0000019  IEA Ensembl:ENSP00000365861 F   Muscleblind-like 2 isoform 1    A3F962_CHICK|MBNL2  protein taxon:9031  20160507    Ensembl
286c286
&lt; UniProtKB A4GTP0  A4GTP0      GO:0044822  GO_REF:0000019  IEA Ensembl:ENSP00000254301 F   Galectin    A4GTP0_CHICK    protein taxon:9031  20160507    Ensembl
<span class="gd">---</span>
&gt; UniProtKB A4GTP0  A4GTP0      GO:0003723  GO_REF:0000019  IEA Ensembl:ENSP00000254301 F   Galectin    A4GTP0_CHICK    protein taxon:9031  20160507    Ensembl
321c321
</pre></div>
</div>
</div>
</div>
<div class="section" id="application-to-human-phenotype-ontology-hpo">
<h2>Application to Human Phenotype Ontology (HPO)<a class="headerlink" href="#application-to-human-phenotype-ontology-hpo" title="Permalink to this headline">¶</a></h2>
<p>Here we show how to use <code class="docutils literal notranslate"><span class="pre">obogaf::parser</span></code> on the <code class="docutils literal notranslate"><span class="pre">HPO</span></code> obo file and its annotation file. Here we go faster, because the experiments are carried-out in the same way of those shown above with the <code class="docutils literal notranslate"><span class="pre">GO</span></code>.</p>
<div class="section" id="parse-the-hpo-obo-file">
<h3>Parse the HPO obo file<a class="headerlink" href="#parse-the-hpo-obo-file" title="Permalink to this headline">¶</a></h3>
<p>Here we use <code class="docutils literal notranslate"><span class="pre">obogaf::parser</span></code> to handle the <code class="docutils literal notranslate"><span class="pre">HPO</span></code> obo file and return some basic statistics. For this example we use the <code class="docutils literal notranslate"><span class="pre">06/09/19</span></code> <code class="docutils literal notranslate"><span class="pre">HPO</span></code> obo release.</p>
<div class="highlight-perl notranslate"><div class="highlight"><pre><span></span><span class="ch">#!/usr/bin/perl</span>

<span class="c1">## loading obogaf::parser and useful Perl module</span>
<span class="k">use</span> <span class="nn">strict</span><span class="p">;</span>
<span class="k">use</span> <span class="nn">warnings</span><span class="p">;</span>
<span class="k">use</span> <span class="nn">File::Path</span> <span class="sx">qw(make_path)</span><span class="p">;</span> <span class="c1">## to recursively create directories</span>
<span class="k">use</span> <span class="nn">obogaf::parser</span><span class="p">;</span>

<span class="c1">## create folder where storing example data</span>
<span class="k">my</span> <span class="nv">$basedir</span><span class="o">=</span> <span class="s">&quot;example/data/&quot;</span><span class="p">;</span>
<span class="n">make_path</span><span class="p">(</span><span class="nv">$basedir</span><span class="p">)</span> <span class="k">unless</span><span class="p">(</span><span class="o">-</span><span class="n">d</span> <span class="nv">$basedir</span><span class="p">);</span>

<span class="c1">## download HPO obo file</span>
<span class="k">my</span> <span class="nv">$obofile</span><span class="o">=</span> <span class="nv">$basedir</span><span class="o">.</span><span class="s">&quot;hpo.obo&quot;</span><span class="p">;</span>
<span class="k">my</span> <span class="nv">$hpobo</span><span class="o">=</span> <span class="sx">qx{wget --output-document=$obofile http://purl.obolibrary.org/obo/hp.obo}</span><span class="p">;</span>
<span class="k">print</span> <span class="s">&quot;HPO obo file downloaded: done\n\n&quot;</span><span class="p">;</span>

<span class="c1">## extract edges from HPO obo file</span>
<span class="k">my</span> <span class="nv">$hpores</span><span class="o">=</span> <span class="nn">obogaf::parser::</span><span class="n">build_edges</span><span class="p">(</span><span class="nv">$obofile</span><span class="p">);</span>
<span class="k">my</span> <span class="nv">$hpoedges</span><span class="o">=</span> <span class="nv">$basedir</span><span class="o">.</span><span class="s">&quot;hpo-edges.txt&quot;</span><span class="p">;</span> <span class="c1">## hpo edges file declared here</span>
<span class="nb">open</span> <span class="n">OUT</span><span class="p">,</span> <span class="s">&quot;&gt; $hpoedges&quot;</span><span class="p">;</span> <span class="c1">## redirect hpo edges on file</span>
<span class="k">print</span> <span class="n">OUT</span> <span class="s">&quot;${$hpores}&quot;</span><span class="p">;</span> <span class="c1">## scalar dereferencing</span>
<span class="nb">close</span> <span class="n">OUT</span><span class="p">;</span>
<span class="k">print</span> <span class="s">&quot;build HPO edges: done\n\n&quot;</span><span class="p">;</span>

<span class="c1">## make stats on HPO</span>
<span class="k">my</span> <span class="p">(</span><span class="nv">$parentIndex</span><span class="p">,</span> <span class="nv">$childIndex</span><span class="p">)</span><span class="o">=</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">);</span>
<span class="k">my</span> <span class="nv">$res</span><span class="o">=</span> <span class="nn">obogaf::parser::</span><span class="n">make_stat</span><span class="p">(</span><span class="nv">$hpoedges</span><span class="p">,</span> <span class="nv">$parentIndex</span><span class="p">,</span> <span class="nv">$childIndex</span><span class="p">);</span>
<span class="k">print</span> <span class="s">&quot;$res&quot;</span><span class="p">;</span> <span class="c1">## print stats on shell</span>

<span class="c1">## results printed on the shell</span>
<span class="c1">#oboterm &lt;tab&gt; degree &lt;tab&gt; indegree &lt;tab&gt; outdegree</span>
<span class="n">HP:0003110</span>  <span class="mi">60</span>  <span class="mi">2</span>   <span class="mi">58</span>
<span class="n">HP:0012379</span>  <span class="mi">45</span>  <span class="mi">1</span>   <span class="mi">44</span>
<span class="n">HP:0010876</span>  <span class="mi">42</span>  <span class="mi">1</span>   <span class="mi">41</span>
<span class="n">HP:0000708</span>  <span class="mi">39</span>  <span class="mi">1</span>   <span class="mi">38</span>
<span class="n">HP:0011805</span>  <span class="mi">39</span>  <span class="mi">1</span>   <span class="mi">38</span>
<span class="n">HP:0003355</span>  <span class="mi">37</span>  <span class="mi">1</span>   <span class="mi">36</span>
<span class="n">HP:0012531</span>  <span class="mi">36</span>  <span class="mi">1</span>   <span class="mi">35</span>
<span class="n">HP:0030057</span>  <span class="mi">34</span>  <span class="mi">1</span>   <span class="mi">33</span>
<span class="n">HP:0001760</span>  <span class="mi">31</span>  <span class="mi">1</span>   <span class="mi">30</span>
<span class="n">HP:0008069</span>  <span class="mi">31</span>  <span class="mi">2</span>   <span class="mi">29</span>
<span class="o">.</span>
<span class="o">.</span>
<span class="o">~</span><span class="n">summary</span> <span class="nb">stat</span><span class="o">~</span>
<span class="n">nodes:</span> <span class="mi">14586</span>
<span class="n">edges:</span> <span class="mi">18416</span>
<span class="n">max</span> <span class="n">degree:</span> <span class="mi">60</span>
<span class="n">min</span> <span class="n">degree:</span> <span class="mi">1</span>
<span class="n">median</span> <span class="n">degree:</span> <span class="mf">1.0000</span>
<span class="n">average</span> <span class="n">degree:</span> <span class="mf">1.2626</span>
<span class="n">density:</span> <span class="mf">8.6567e-05</span>
</pre></div>
</div>
</div>
<div class="section" id="parse-the-hpo-annotation-file">
<h3>Parse the HPO annotation file<a class="headerlink" href="#parse-the-hpo-annotation-file" title="Permalink to this headline">¶</a></h3>
<p>Here we use <code class="docutils literal notranslate"><span class="pre">obogaf::parser</span></code> to parse the <code class="docutils literal notranslate"><span class="pre">HPO</span></code> annotation file (release <code class="docutils literal notranslate"><span class="pre">06/09/19</span></code>)</p>
<div class="highlight-perl notranslate"><div class="highlight"><pre><span></span><span class="ch">#!/usr/bin/perl</span>

<span class="c1">## loading obogaf::parser and useful Perl module</span>
<span class="k">use</span> <span class="nn">strict</span><span class="p">;</span>
<span class="k">use</span> <span class="nn">warnings</span><span class="p">;</span>
<span class="k">use</span> <span class="nn">File::Path</span> <span class="sx">qw(make_path)</span><span class="p">;</span> <span class="c1">## to recursively create directories</span>
<span class="k">use</span> <span class="nn">obogaf::parser</span><span class="p">;</span>

<span class="c1">## create folder where storing data</span>
<span class="k">my</span> <span class="nv">$basedir</span><span class="o">=</span> <span class="s">&quot;example/data/&quot;</span><span class="p">;</span>
<span class="n">make_path</span><span class="p">(</span><span class="nv">$basedir</span><span class="p">)</span> <span class="k">unless</span><span class="p">(</span><span class="o">-</span><span class="n">d</span> <span class="nv">$basedir</span><span class="p">);</span>

<span class="c1">## download HPO annotations</span>
<span class="k">my</span> <span class="nv">$hpofile</span><span class="o">=</span> <span class="nv">$basedir</span><span class="o">.</span><span class="s">&quot;hpo.ann.txt&quot;</span><span class="p">;</span> <span class="c1">## hpo annotation file declared here</span>
<span class="k">my</span> <span class="nv">$hpoann</span><span class="o">=</span> <span class="sx">qx{wget --output-document=$hpofile http://compbio.charite.de/jenkins/job/hpo.annotations.monthly/lastStableBuild/artifact/annotation/ALL_SOURCES_ALL_FREQUENCIES_genes_to_phenotype.txt}</span><span class="p">;</span>

<span class="c1">## extract HPO annotations</span>
<span class="k">my</span> <span class="p">(</span><span class="nv">$geneindex</span><span class="p">,</span> <span class="nv">$classindex</span><span class="p">)</span><span class="o">=</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">3</span><span class="p">);</span>
<span class="k">my</span> <span class="p">(</span><span class="nv">$res</span><span class="p">,</span> <span class="nv">$stat</span><span class="p">)</span><span class="o">=</span> <span class="nn">obogaf::parser::</span><span class="n">gene2biofun</span><span class="p">(</span><span class="nv">$hpofile</span><span class="p">,</span> <span class="nv">$geneindex</span><span class="p">,</span> <span class="nv">$classindex</span><span class="p">);</span>
<span class="k">my</span> <span class="nv">$hpout</span><span class="o">=</span> <span class="nv">$basedir</span><span class="o">.</span><span class="s">&quot;hpo.gene2pheno.txt&quot;</span><span class="p">;</span> <span class="c1">## annotation adj list stored in a file</span>
<span class="nb">open</span> <span class="n">OUT</span><span class="p">,</span> <span class="s">&quot;&gt; $hpout&quot;</span><span class="p">;</span>
<span class="k">foreach</span> <span class="k">my</span> <span class="nv">$k</span> <span class="p">(</span><span class="nb">sort</span><span class="p">{</span><span class="nv">$a</span> <span class="ow">cmp</span> <span class="nv">$b</span><span class="p">}</span> <span class="nb">keys</span> <span class="nv">%$res</span><span class="p">)</span> <span class="p">{</span> <span class="k">print</span> <span class="n">OUT</span> <span class="s">&quot;$k $$res{$k}\n&quot;</span><span class="p">;}</span> <span class="c1">## dereferencing</span>
<span class="nb">close</span> <span class="n">OUT</span><span class="p">;</span>
<span class="k">print</span> <span class="s">&quot;${$stat}\n&quot;</span><span class="p">;</span>

<span class="c1">## results printed on the shell</span>
<span class="n">genes:</span> <span class="mi">4231</span>
<span class="n">ontology</span> <span class="n">terms:</span> <span class="mi">7616</span>
</pre></div>
</div>
<p>Below we show the first <code class="docutils literal notranslate"><span class="pre">10</span></code> lines of the <code class="docutils literal notranslate"><span class="pre">hpo.gene2pheno.txt</span></code> file, just to give an example of how this file is structured:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>A2M HP:0410054|HP:0001425|HP:0001300|HP:0000006|HP:0000726|HP:0002423|HP:0002185|HP:0002511
A2ML1 HP:0000768|HP:0001156|HP:0000006|HP:0000391|HP:0001928|HP:0000520|HP:0100625|HP:0000403|HP:0000407|HP:0011800|HP:0011675|HP:0000028|HP:0002974|HP:0002208|HP:0008872|HP:0000044|HP:0001324|HP:0000179|HP:0007477|HP:0005692|HP:0000316|HP:0002750|HP:0004415|HP:0002240|HP:0000325|HP:0010318|HP:0001743|HP:0000465|HP:0006610|HP:0002650|HP:0000218|HP:0000474|HP:0000347|HP:0000348|HP:0000476|HP:0011869|HP:0004322|HP:0011362|HP:0000995|HP:0001252|HP:0001892|HP:0000486|HP:0010982|HP:0001641|HP:0001260|HP:0001004|HP:0000494|HP:0000368|HP:0004209|HP:0002162|HP:0011381|HP:0000508|HP:0000639|HP:0000767
A4GALT HP:0000006|HP:0010970
AAAS HP:0001347|HP:0008259|HP:0007556|HP:0000007|HP:0011463|HP:0002376|HP:0000648|HP:0000649|HP:0000522|HP:0002571|HP:0000972|HP:0000846|HP:0007440|HP:0001430|HP:0000982|HP:0000407|HP:0007002|HP:0003676|HP:0003487|HP:0004319|HP:0001761|HP:0001249|HP:0001250|HP:0004322|HP:0001251|HP:0008163|HP:0001252|HP:0000612|HP:0012332|HP:0001260|HP:0001324|HP:0002093|HP:0001263|HP:0010486|HP:0000505|HP:0000953|HP:0000252|HP:0009916|HP:0000830|HP:0001278
AAGAB HP:0003584|HP:0040162|HP:0025092|HP:0000006|HP:0007530|HP:0002894|HP:0005584|HP:0001425|HP:0006740|HP:0000982|HP:0025114|HP:0003002|HP:0003003|HP:0001597|HP:0012189
AARS1 HP:0003202|HP:0000643|HP:0001284|HP:0000006|HP:0000007|HP:0000648|HP:0001290|HP:0002827|HP:0002059|HP:0002317|HP:0002063|HP:0001298|HP:0003477|HP:0001558|HP:0000407|HP:0002072|HP:0002460|HP:0000668|HP:0012447|HP:0000546|HP:0002355|HP:0100660|HP:0001336|HP:0001337|HP:0011968|HP:0009027|HP:0200134|HP:0002376|HP:0000717|HP:0002509|HP:0002133|HP:0002521|HP:0000348|HP:0010844|HP:0001761|HP:0001249|HP:0001250|HP:0004322|HP:0001251|HP:0001508|HP:0002020|HP:0001765|HP:0003429|HP:0009830|HP:0003431|HP:0001511|HP:0100710|HP:0001257|HP:0007018|HP:0000494|HP:0000750|HP:0001263|HP:0001265|HP:0003828|HP:0001268|HP:0002421|HP:0002936|HP:0000504|HP:0003577|HP:0001273|HP:0000252|HP:0000508|HP:0000639
AARS2 HP:0002371|HP:0002180|HP:0006980|HP:0000007|HP:0002186|HP:0000716|HP:0008209|HP:0000726|HP:0003676|HP:0001251|HP:0001508|HP:0001639|HP:0002151|HP:0001257|HP:0002089|HP:0001260|HP:0002353|HP:0001522|HP:0001332|HP:0001272|HP:0003128|HP:0001337|HP:0006970|HP:0003324|HP:0000639
AASS HP:0000736|HP:0001249|HP:0003297|HP:0004322|HP:0001250|HP:0001252|HP:0000007|HP:0001256|HP:0003593|HP:0000750|HP:0002927|HP:0001903|HP:0001264|HP:0000752|HP:0002353|HP:0002161|HP:0000119|HP:0001083|HP:0100543
ABAT HP:0001250|HP:0000098|HP:0001347|HP:0001254|HP:0000007|HP:0001321|HP:0003819|HP:0025356|HP:0006829|HP:0000494|HP:0001263|HP:0002415|HP:0025430|HP:0000278|HP:0001274|HP:0007291
ABCA1 HP:0002240|HP:0003457|HP:0100546|HP:0003396|HP:0001349|HP:0000006|HP:0000007|HP:0025608|HP:0003146|HP:0010829|HP:0001677|HP:0004943|HP:0007759|HP:0000656|HP:0001744|HP:0001873|HP:0008404|HP:0003477|HP:0007957|HP:0004374|HP:0011096|HP:0005145|HP:0001433|HP:0002460|HP:0002716|HP:0007133|HP:0030814|HP:0000991|HP:0007328|HP:0003233|HP:0002730|HP:0002027|HP:0002155|HP:0003693|HP:0000622|HP:0001903|HP:0001712|HP:0001392|HP:0001265|HP:0002164|HP:0006901|HP:0000505|HP:0001658|HP:0005181|HP:0000958

... to be continued ...
</pre></div>
</div>
</div>
<div class="section" id="map-hpo-terms-between-releases">
<h3>Map HPO terms between releases<a class="headerlink" href="#map-hpo-terms-between-releases" title="Permalink to this headline">¶</a></h3>
<p>Here we use <code class="docutils literal notranslate"><span class="pre">obogaf::parser</span></code> to map the <code class="docutils literal notranslate"><span class="pre">HPO</span></code> terms of an <em>old</em> release (<code class="docutils literal notranslate"><span class="pre">09/03/18</span></code>) toward a <em>new</em> ones (<code class="docutils literal notranslate"><span class="pre">06/09/19</span></code>).</p>
<div class="highlight-perl notranslate"><div class="highlight"><pre><span></span><span class="ch">#!/usr/bin/perl</span>

<span class="c1">## loading obogaf::parser and useful Perl module</span>
<span class="k">use</span> <span class="nn">strict</span><span class="p">;</span>
<span class="k">use</span> <span class="nn">warnings</span><span class="p">;</span>
<span class="k">use</span> <span class="nn">File::Path</span> <span class="sx">qw(make_path)</span><span class="p">;</span> <span class="c1">## to recursively create directories</span>
<span class="k">use</span> <span class="nn">obogaf::parser</span><span class="p">;</span>

<span class="c1">## create folder where storing data</span>
<span class="k">my</span> <span class="nv">$basedir</span><span class="o">=</span> <span class="s">&quot;example/data/&quot;</span><span class="p">;</span>
<span class="n">make_path</span><span class="p">(</span><span class="nv">$basedir</span><span class="p">)</span> <span class="k">unless</span><span class="p">(</span><span class="o">-</span><span class="n">d</span> <span class="nv">$basedir</span><span class="p">);</span>

<span class="c1">## download HPO obo file</span>
<span class="k">my</span> <span class="nv">$obofile</span><span class="o">=</span> <span class="nv">$basedir</span><span class="o">.</span><span class="s">&quot;hpo.obo&quot;</span><span class="p">;</span>
<span class="k">my</span> <span class="nv">$hpobo</span><span class="o">=</span> <span class="sx">qx{wget --output-document=$obofile http://purl.obolibrary.org/obo/hp.obo}</span><span class="p">;</span>

<span class="c1">## download HPO old annotation file</span>
<span class="k">my</span> <span class="nv">$hpofileOld</span><span class="o">=</span> <span class="nv">$basedir</span><span class="o">.</span><span class="s">&quot;hpo.ann.old.txt&quot;</span><span class="p">;</span> <span class="c1">## goa annotation file declared here</span>
<span class="k">my</span> <span class="nv">$hpold</span><span class="o">=</span> <span class="sx">qx{wget --output-document=$hpofileOld http://compbio.charite.de/jenkins/job/hpo.annotations.monthly/139/artifact/annotation/ALL_SOURCES_ALL_FREQUENCIES_genes_to_phenotype.txt}</span><span class="p">;</span>

<span class="c1">## map HPO terms between releases</span>
<span class="k">my</span> <span class="nv">$classindex</span><span class="o">=</span> <span class="mi">3</span><span class="p">;</span>
<span class="k">my</span> <span class="p">(</span><span class="nv">$res</span><span class="p">,</span> <span class="nv">$stat</span><span class="p">)</span><span class="o">=</span> <span class="nn">obogaf::parser::</span><span class="n">map_OBOterm_between_release</span><span class="p">(</span><span class="nv">$obofile</span><span class="p">,</span> <span class="nv">$hpofileOld</span><span class="p">,</span> <span class="nv">$classindex</span><span class="p">);</span>
<span class="k">my</span> <span class="nv">$mapfile</span><span class="o">=</span> <span class="nv">$basedir</span><span class="o">.</span><span class="s">&quot;hpo.ann.mapped.txt&quot;</span><span class="p">;</span>
<span class="nb">open</span> <span class="n">OUT</span><span class="p">,</span> <span class="s">&quot;&gt; $mapfile&quot;</span><span class="p">;</span> <span class="c1">## mapped annotation stored in a file</span>
<span class="k">print</span> <span class="n">OUT</span> <span class="s">&quot;${$res}&quot;</span><span class="p">;</span>
<span class="nb">close</span> <span class="n">OUT</span><span class="p">;</span>
<span class="k">print</span> <span class="s">&quot;${$stat}&quot;</span><span class="p">;</span>

<span class="c1">## results printed on shell</span>
<span class="c1">#alt-id &lt;tab&gt; id</span>
<span class="n">HP:0000487</span>  <span class="n">HP:0000486</span>
<span class="n">HP:0000547</span>  <span class="n">HP:0000510</span>
<span class="n">HP:0000655</span>  <span class="n">HP:0007773</span>
<span class="n">HP:0000833</span>  <span class="n">HP:0001952</span>
<span class="n">HP:0001226</span>  <span class="n">HP:0006121</span>
<span class="n">HP:0001322</span>  <span class="n">HP:0006872</span>
<span class="n">HP:0001472</span>  <span class="n">HP:0001426</span>
<span class="n">HP:0001862</span>  <span class="n">HP:0006121</span>
<span class="n">HP:0002271</span>  <span class="n">HP:0012332</span>
<span class="n">HP:0002281</span>  <span class="n">HP:0002282</span>
<span class="n">HP:0002459</span>  <span class="n">HP:0012332</span>
<span class="n">HP:0003464</span>  <span class="n">HP:0003107</span>
<span class="n">HP:0003490</span>  <span class="n">HP:0003150</span>
<span class="n">HP:0005130</span>  <span class="n">HP:0001723</span>
<span class="n">HP:0005364</span>  <span class="n">HP:0004429</span>
<span class="n">HP:0005901</span>  <span class="n">HP:0002754</span>
<span class="n">HP:0006830</span>  <span class="n">HP:0001319</span>
<span class="n">HP:0007314</span>  <span class="n">HP:0002282</span>
<span class="n">HP:0007519</span>  <span class="n">HP:0007485</span>
<span class="n">HP:0007713</span>  <span class="n">HP:0010920</span>
<span class="n">HP:0007758</span>  <span class="n">HP:0000505</span>
<span class="n">HP:0007868</span>  <span class="n">HP:0000608</span>
<span class="n">HP:0007893</span>  <span class="n">HP:0000546</span>
<span class="n">HP:0008012</span>  <span class="n">HP:0000545</span>
<span class="n">HP:0008024</span>  <span class="n">HP:0100018</span>
<span class="n">HP:0008230</span>  <span class="n">HP:0040171</span>
<span class="n">HP:0010700</span>  <span class="n">HP:0000518</span>
<span class="n">HP:0011146</span>  <span class="n">HP:0002384</span>
<span class="n">HP:0012201</span>  <span class="n">HP:0008151</span>
<span class="n">HP:0040290</span>  <span class="n">HP:0003011</span>
<span class="n">HP:0045016</span>  <span class="n">HP:0003455</span>

<span class="n">Tot</span><span class="o">.</span> <span class="n">ontology</span> <span class="n">terms:</span>    <span class="mi">6789</span>
<span class="n">Tot</span><span class="o">.</span> <span class="n">altID:</span> <span class="mi">3634</span>
<span class="n">Tot</span><span class="o">.</span> <span class="n">altID</span> <span class="n">seen:</span>    <span class="mi">31</span>
<span class="n">Tot</span><span class="o">.</span> <span class="n">altID</span> <span class="n">unseen:</span>  <span class="mi">3603</span>
</pre></div>
</div>
<p>By running the <code class="docutils literal notranslate"><span class="pre">diff</span></code> command between the <em>old</em> file (<code class="docutils literal notranslate"><span class="pre">hpo.ann.old.txt</span></code>) and the <em>mapped</em> one (<code class="docutils literal notranslate"><span class="pre">hpo.ann.mapped.txt</span></code>) and redirecting the results on a output file (e.g.: <code class="docutils literal notranslate"><span class="pre">diff</span> <span class="pre">hpo.ann.old.txt</span> <span class="pre">hpo.ann.mapped.txt</span> <span class="pre">&gt;</span> <span class="pre">hpo.ann.diff</span></code>) we can easily visualize the changed <code class="docutils literal notranslate"><span class="pre">HPO</span></code> terms between the two release. Below  we show just some few lines of <code class="docutils literal notranslate"><span class="pre">hpo.ann.diff</span></code> to give an example:</p>
<div class="highlight-diff notranslate"><div class="highlight"><pre><span></span>1148c1148
&lt; 51    ACOX1   Tapetoretinal degeneration  HP:0000547
<span class="gd">---</span>
&gt; 51    ACOX1   Tapetoretinal degeneration  HP:0000510
3423c3423
&lt; 190   NR0B1   Decreased testosterone in males HP:0008230
<span class="gd">---</span>
&gt; 190   NR0B1   Decreased testosterone in males HP:0040171
4041c4041
&lt; 212   ALAS2   Glucose intolerance HP:0000833
<span class="gd">---</span>
&gt; 212   ALAS2   Glucose intolerance HP:0001952
5049c5049
&lt; 8481  OFD1    Gray matter heterotopias    HP:0002281
<span class="gd">---</span>
&gt; 8481  OFD1    Gray matter heterotopias    HP:0002282
6597c6597
&lt; 57724 EPG5    White matter neuronal heterotopia   HP:0007314
<span class="gd">---</span>
&gt; 57724 EPG5    White matter neuronal heterotopia   HP:0002282
7244c7244
&lt; 429   ASCL1   Dysautonomia    HP:0002459
<span class="gd">---</span>
&gt; 429   ASCL1   Dysautonomia    HP:0012332
7246c7246
</pre></div>
</div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="faq.html" class="btn btn-neutral float-right" title="Frequently Asked Questions" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="usage.html" class="btn btn-neutral float-left" title="Usage of obogaf::parser" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, Marco Notaro

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>